File: problems/15_greedy/0763_partition_labels/solution.swift
// 0763. Partition Labels
import Foundation
class Solution {
    func partitionLabels(_ s: String) -> [Int] {
        let a = Array(s)
        var last = [Character: Int]()
        for i in 0..<a.count { last[a[i]] = i }
        var res = [Int]()
        var start = 0, end = 0
        for i in 0..<a.count {
            if let v = last[a[i]], v > end { end = v }
            if i == end { res.append(end - start + 1); start = i + 1 }
        }
        return res
    }
}
let sol = Solution()
if sol.partitionLabels("ababcbacadefegdehijhklij") != [9, 7, 8] { fatalError("a") }
if sol.partitionLabels("eccbbbbdec") != [10] { fatalError("b") }
if sol.partitionLabels("a") != [1] { fatalError("c") }
if sol.partitionLabels("abc") != [1, 1, 1] { fatalError("d") }
if sol.partitionLabels("aaabbb") != [3, 3] { fatalError("e") }
print("All tests passed!")

File: problems/15_greedy/0763_partition_labels/solution.py
"""
Problem: 0763. Partition Labels
"""
from typing import List


class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        last = {c: i for i, c in enumerate(s)}
        res = []
        start = end = 0
        for i, c in enumerate(s):
            if last[c] > end:
                end = last[c]
            if i == end:
                res.append(end - start + 1)
                start = i + 1
        return res


if __name__ == "__main__":
    sol = Solution()
    assert sol.partitionLabels("ababcbacadefegdehijhklij") == [9, 7, 8]
    assert sol.partitionLabels("eccbbbbdec") == [10]
    assert sol.partitionLabels("a") == [1]
    assert sol.partitionLabels("abc") == [1, 1, 1]
    assert sol.partitionLabels("aaabbb") == [3, 3]
    print("All tests passed!")
